designs/2023-prelints/README.md

```diff
@@ -0,0 +1,452 @@
+     * The name of the file without a path.
+     */
+    filename: string;
+
+    /**
+     * The body of the file to parse.
+     */
+    body: ArrayBuffer;
+
+    /**
+     * The number to add to violation offset locations. 0 if undefined.
+     */
+    byteStart?: number;
+
+    /**
+     * The number to add to the offset location of each violation in the body.
+     * 0 if undefined.
+     */
+    byteOffset?: number;
+}
+
+interface SourceCode {
+    // language-defined
+}
+```
+
+Each prelint follows the same basic structure of as a rule, only it cannot report violations and instead can only inspect the AST, use `SourceCode` to make any necessary changes, and create virtual files as necessary. Otherwise, it looks the same as a rule, such as:
+
+```js
+export default {
+    meta: {
+        description: "My prelint"
+    },
+    create(context) {
+
+        const sourceCode = context.sourceCode;
+
+        // work with sourceCode here
+
+        return {
+            Identifier() {
+                // visitor method
+            }
+        }
+    }
+};
+```
+
+Because the prelints follow the same format as rules, we can use the same type of traversal.
+
+#### The `PrelintContext#createVirtualFile()` method
+
+The design of `VirtualTextFile` is intended to automate the mapping of line/column reporting that currently has to be done manually inside of a processor in the `postprocess()` step. By specifying the line and column location of the virtual file inside its parent, we can calculate that automatically.
+
+During prelint traversal, the core will gather up all of the created virtual files and then queue them up for linting in addition to the original (parent) file. To each rule, virtual files would look the same as if they came from a processor.
+
+This could make processors obsolete going forward.
+
+As an example, suppose we have an HTML language plugin that can parse the following file:
+
+```html
+<!doctype html>
+<html>
+<head>
+    <title>Example</title>
+    <script>
+        function sayHi() {
+            console.log("Hi");
+        }
+    </script>
+    <style>
+        body {
+            background: blue;
+        }
+    </style>
+</head>
+</html>
+```
+
+The HTML language plugin will first parse the HTML into an AST. Inside of that AST, there will be a `Script` node representing the JavaScript portion of the file and a `Style` node representing the CSS portion. We could create a virtual file for each use prelints. Here is an example that extracts the JavaScript into a virtual file:
+
+```js
+export default {
+    meta: {
+        description: "Extracts JS from HTML."
+    },
+    create(context) {
+
+        const sourceCode = context.sourceCode;
+
+        return {
+            Script(node) {
+
+                // imagine the content property exists
+                const textNode = node.content;
+
+                context.createVirtualFile({
+                    filename: "0.js",
+                    body: removeIndents(
```

Comment:
Slightly off-topic - I would envision that most tools would want to strip indent in the same way - might be good to provide a util from eslint itself that does this?

Reply:
Certainly worth thinking about -- also worth thinking about: does ESLint just do it automatically for you based on the `indentOffset`?

Reply:
Updated the proposal to automatically adjust indents.